src/recommendation/src/main/scala/RankingTrainValidationSplit.scala

```diff
@@ -0,0 +1,343 @@
+        }
+      case a: ALS =>
+        $(estimator).asInstanceOf[ALS]
+    }
+
+    val userColumn = est.getUserCol
+    val itemColumn = est.getItemCol
+
+    val perUserRecommendedItemsDF: DataFrame = recs
+      .select(userColumn, "recommendations." + itemColumn)
+      .withColumnRenamed(itemColumn, "prediction")
+
+    val perUserActualItemsDF = if (validationDataset.columns.contains($(ratingCol))) {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(ratingCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn, $(ratingCol))
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
+    } else {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(itemCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn)
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
+    }
+    val joined_rec_actual = perUserRecommendedItemsDF
+      .join(perUserActualItemsDF, userColumn)
+      .drop(userColumn)
+
+    joined_rec_actual
+  }
+}
+
+object RankingTrainValidationSplit extends ComplexParamsReadable[RankingTrainValidationSplit] {
+  def popRow(r: Row): Any = r.getDouble(1)
+}
+
+@InternalWrapper
+class RankingTrainValidationSplitModel(
+  override val uid: String)
+  extends Model[RankingTrainValidationSplitModel] with RankingTrainValidationSplitParams
+    with ComplexParamsWritable {
+
+  def setValidationMetrics(value: Array[_]): this.type = set(validationMetrics, value)
+
+  val validationMetrics = new ArrayParam(this, "validationMetrics", "Best Model")
+
+  /** @group getParam */
+  def getValidationMetrics: Array[_] = $(validationMetrics)
+
+  def setBestModel(value: Model[_]): this.type = set(bestModel, value)
+
+  val bestModel: TransformerParam =
+    new TransformerParam(
+    this,
+    "bestModel", "The internal ALS model used splitter",
+    { t => t.isInstanceOf[Model[_]] })
+
+  /** @group getParam */
+  def getBestModel: Model[_] = $(bestModel).asInstanceOf[Model[_]]
+
+  def this() = this(Identifiable.randomUID("RankingTrainValidationSplitModel"))
+
+  override def copy(extra: ParamMap): RankingTrainValidationSplitModel = {
+    val copied = new RankingTrainValidationSplitModel(uid)
+    copyValues(copied, extra).setParent(parent)
+  }
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+
+    //sort to pass unit test
+    $(bestModel).transform(dataset).sort("prediction")
```

Comment:
can you explain this a bit more, why do you need the sort here?

Reply:
in order to pass a unit test.... if I don't sort the matching dataframe unit test fails.